### PR TITLE
Timeout command created

### LINKED ISSUE #468: Add a timeout command
**Is your feature request related to a problem? Please describe.**

Currently, the bot has no command that utilizes Discord's recent [Time Out feature](https://support.discord.com/hc/en-us/articles/4413305239191). This means that in any servers utilizing CostBot, if mods want to time out a member, they're unable to do so using the bot and have to either resort to using another bot or doing it through the UI. 

**Describe the solution you'd like**

Implement a `timeout` command. The command would be split into two subcommands, one of them being `add` and the other being `remove`.

The `add` subcommand would first check the member's permissions, since members with the `ADMINISTRATOR` permission, as well as server owners, cannot be timed out. Afterwards, it would time out the user, and depending on the options supplied, inform them of the duration & reason for the time out, based on the options provided by the moderator. 

The `remove` subcommand would simply allow moderators to remove any ongoing timeouts.

**Describe alternatives you've considered**

There are no alternative solutions, other than not implementing the feature at all.

**Additional context**

### PR DESCRIPTION
**Pull request by Ladam0203**
### **Please describe all changes made by this Pull Request and why you feel like it should be merged:**

I have created the timeout command. It is split into two subcommands: add and remove. 

Features checks in a similiar fashion as it is done in other commands of this bot:
- The bot's and also the executing member's permission to ModerateMembers is checked. 
- It is check whether the supplied user is valid - however this might be unnecessary, since the UserOption only accepts valid users - but it was done like that in other commands, so I went for consistency
- Issuers of this command cannot time out themselves
- Admins and the server owner cannot be timed out.

Does not feature a DM, as the timeout reason and time can be directly seen by the user, but can be added if requested.

### **Issues**

This PR closes #468.

### **Checklists**

-   [x] I have read and agreed to both the [Code of Conduct](https://github.com/Costpap/CostBot/blob/main/.github/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/Costpap/CostBot/blob/main/.github/CONTRIBUTING.md) (**required**)
-   [x] Code changes have been tested **with a bot account on Discord**, or there are none. (**required**)
-   [x] Any code changes have been checked against ESLint (`npx eslint .`) and the TypeScript compiler (`npx tsc`) and there are no errors, or there are no code changes. (**required**)

### CODE REVIEW

**Comment by Ladam0203** [comment]
Hmm... not sure why the lint is not passing... I ran the test and the fix as you have requested... let me try again.

**Comment by Ladam0203** [comment]
Please let me know if I need to do something additional, e.g. set up a GPG key.

**Comment by Ladam0203** [comment]
Added the GPG signature as well, just to be sure. How everything checks out now :)

**Comment by Costpap** [comment]
As promised in #468, [here's the timeout command I was working on](https://gist.github.com/Costpap/c23f57ac41a590e86ebb31379691439c). It's not entirely finished and it does need a few improvements, however, if you see anything you feel like could be worth adding to your iteration of the command, feel free to do so!

I'll go ahead and review the pull request once you decide whether or not you want to make any changes to your current iteration of the command.

**Comment by Ladam0203** [comment]
Hey! I have finished going through your gist... I made some changes to my version based on that, keeping mind consistency. 

I went with the ms lib in the end for the duration, however I provided some example in the decription of the option, so it is easier for the issuer to figure out what to type
I added extra checks where it was necessary.
I re-checked spelling and grammar.
I use the notifyUser method to send a notification about a timeout to a user.

And I am totally forgetting something, but you will see it in the code... 

Let me know what do you think!

File: src/commands/timeout.ts [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+
+//imports
+import { ChatInputCommandInteraction, PermissionFlagsBits, SlashCommandBuilder } from 'discord.js';
+import { notifyUser } from '../utils/misc';
+import ms from 'ms';
+
+export default {
+    data: new SlashCommandBuilder()
+        .setName('timeout')
+        .setDescription('Manages the timeout of a selected user.')
```
Comment (Costpap) [inline src/commands/timeout.ts:11]: Please change the string here to be consistent with the one initially written in the gist.
```suggestion
        .setDescription('Manages user timeouts.')
```

Comment (Costpap) [inline src/commands/timeout.ts:11]: Please change the string here to be consistent with the one initially written in the gist.
```suggestion
        .setDescription('Manages user timeouts.')
```

File: src/commands/timeout.ts [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+                    option.setName('reason').setDescription('Reason for timing out the user').setRequired(false),
+                )
+                .addBooleanOption((option) =>
+                    option.setName('silent').setDescription('Whether or not to DM the user').setRequired(false),
+                ),
+        )
+        .addSubcommand((subcommand) =>
+            subcommand
+                .setName('remove')
+                .setDescription('Removes timeout from a user.')
```
Comment (Costpap) [inline src/commands/timeout.ts:33]: An `a` is missing before `timeout`.
```suggestion
                .setDescription('Removes a timeout from a user.')
```

File: src/commands/timeout.ts [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+        // Typeguard in order to ensure having access to ChatInputCommand interaction options.
+        if (!interaction.isChatInputCommand()) return;
+
+        if (interaction.inGuild() === false) {
+            return interaction.reply({ content: "❌ I can't execute this command inside DMs!", ephemeral: true });
+        }
+
+        if (!interaction.guild.members.me.permissions.has(PermissionFlagsBits.ModerateMembers)) {
+            return interaction.reply({
+                content: '❌ Sorry, I need the `Moderate Members` in order to execute this command.',
```
Comment (Costpap) [inline src/commands/timeout.ts:51]: Please change `Moderate Members` to `Timeout Members`, in accordance with what is shown in Discord's UI.
```suggestion
                content: '❌ Sorry, I need the `Moderate Members` in order to execute this command.',
```

File: src/commands/timeout.ts [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+                ephemeral: true,
+            });
+        }
+        if (typeof interaction.member.permissions === 'string')
+            return interaction.reply({ content: '❌ Unknown permissions. Please try again later.', ephemeral: true });
+        else if (!interaction.member.permissions.has(PermissionFlagsBits.ModerateMembers)) {
+            return interaction.reply({
+                content: '⛔ You need the `Moderate Members` permission in order to use this command!',
```
Comment (Costpap) [inline src/commands/timeout.ts:59]: Please change `Moderate Members` to `Timeout Members`, in accordance with what is shown in Discord's UI.
```suggestion
                content: '⛔ You need the `Timeout Members` permission in order to use this command!',
```

File: src/commands/timeout.ts [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+            if (msDuration == null) {
+                return interaction.reply({
+                    content: '❌ The format of the timeout duration is invalid!',
+                    ephemeral: true,
+                });
+            }
+
+            if (msDuration <= 0) {
+                return interaction.reply({
+                    content: '❌ The duration of the timeout must be greater than 0!',
```
Comment (Costpap) [inline src/commands/timeout.ts:99]: I'd rather see this string changed to the one below.
```suggestion
                    content: '❌ Please specify a valid timeout duration.',
```

File: src/commands/timeout.ts [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+
+            const duration = interaction.options.getString('duration', true);
+            const msDuration = ms(duration);
+
+            if (msDuration == null) {
+                return interaction.reply({
+                    content: '❌ The format of the timeout duration is invalid!',
+                    ephemeral: true,
+                });
+            }
```
Comment (Costpap) [inline src/commands/timeout.ts:95]: I feel like this wouldn't work out. Could you please show me what the bot sends when you try passing a duration that can't get parsed into a number on Discord?

File: src/commands/timeout.ts [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+
+            if (msDuration <= 0) {
+                return interaction.reply({
+                    content: '❌ The duration of the timeout must be greater than 0!',
+                    ephemeral: true,
+                });
+            }
+
+            let response = `⏳ Timed out \`${member.user.tag} (${member.id})\`.`;
```
Comment (Costpap) [inline src/commands/timeout.ts:104]: I'd rather see the ✅ emoji get used instead of ⌛.
```suggestion
            let response = `✅ Timed out \`${member.user.tag} (${member.id})\`.`;
```

**Review by Costpap — CHANGES_REQUESTED** [review]
After having a quick look at all of the code you've written, I've pointed out a few changes that could be made to the strings used. 

In addition, the version of the command I came up with used a switch statement to run the different code for each subcommand. From what I see, your code doesn't include a switch statement. Could you please change the two if statements to a single switch statement? In my opinion, it is cleaner and more consistent with the rest of the codebase, namely the bot's `/legal` command.

That's all for now, let me know when all of the changes requested are implemented. I'll have a more thorough look then, and preferably will spin up an instance of the bot with the command to test it out locally.

File: src/commands/timeout.ts (hunk 2 above)
Comment (Ladam0203) [reply src/commands/timeout.ts:33]: I feel like the "a" is unnecessary, as a user can only have one timeout.

File: src/commands/timeout.ts (hunk 3 above)
Comment (Ladam0203) [reply src/commands/timeout.ts:51]: I can make this change, however, I think it is shown as "Moderate Members" even on the UI. Photo attached. 
image

File: src/commands/timeout.ts (hunk 4 above)
Comment (Ladam0203) [reply src/commands/timeout.ts:59]: Same as prevoiusly written.
image

File: src/commands/timeout.ts (hunk 6 above)
Comment (Ladam0203) [reply src/commands/timeout.ts:95]: I have tested this, and the ms() method returns undefined when a string cannot be parsed. I used loosely equals null, because that checks for undefined (if the string cannot be parsed) and also for null (if the duration would be missing for some reason)

I can change it to === undefined, but I think this is the safest option

File: src/commands/timeout.ts (hunk 5 above)
Comment (Ladam0203) [reply src/commands/timeout.ts:99]: Done, however, wouldn't you like an exclamation mark at the end? As it is as error message.

File: src/commands/timeout.ts (hunk 1 above)
Comment (Ladam0203) [reply src/commands/timeout.ts:11]: Done.

File: src/commands/timeout.ts (hunk 3 above)
Comment (Costpap) [reply src/commands/timeout.ts:51]: Where did you get that screenshot from? Normally, when using the en-US locale, the permission tied to timing out members is "Timeout Members". cc my image below.

image

File: src/commands/timeout.ts (hunk 5 above)
Comment (Costpap) [reply src/commands/timeout.ts:99]: Not all error messages have exclamation marks at the end, whether or not to include one is decided by how it sounds with or without one when reading the error message out loud. I don't see a need for a question mark here, personally.

File: src/commands/timeout.ts (hunk 6 above)
Comment (Costpap) [reply src/commands/timeout.ts:95]: Hmm. I'll have to experiment with this myself a little bit, if you don't mind. :)

**Comment by Costpap** [comment]
Things are looking better, the main thing I'm concerned with is that I may not have enough time to fully review, test and merge the pull request until Friday comes around. Totally hope you don't mind that. At the moment, I don't see anything else that could require changes aside from the things I mentioned in my review comment, but I could be totally wrong due to not having enough time to take a look at every single line of code you've written.

File: src/commands/timeout.ts (hunk 3 above)
Comment (Ladam0203) [reply src/commands/timeout.ts:51]: This is from the English, UK client. However, I think it will be later changed to Timeout Members there as well, as it would makes sense. Maybe if we are lucky, thel will change the PermissionBitField to TimeoutMembers as well in the future.

I'll go with Timeout Members in the text.

File: src/commands/timeout.ts (hunk 5 above)
Comment (Ladam0203) [reply src/commands/timeout.ts:99]: Okay, I'll go with the . instead of an !

File: src/commands/timeout.ts (hunk 6 above)
Comment (Ladam0203) [reply src/commands/timeout.ts:95]: Let me know what you found! :)

File: src/commands/timeout.ts (hunk 7 above)
Comment (Ladam0203) [reply src/commands/timeout.ts:104]: I have made this change

**Comment by Ladam0203** [comment]
Okay, let me know what should happen with the "An a is missing before timeout." and the ms parsing == null thingy.

After those, I'll make a new commit with your requested changes at Thursday evening, so you can review it on Friday :)

No worries, I have only opened this PR bc I need points for HackSquad, but my team already abandoned the competition and I do not have time for it either afer all, so it is a chill pace :D

File: src/commands/timeout.ts (hunk 2 above)
Comment (Costpap) [reply src/commands/timeout.ts:33]: Thing is, without the "a", it would be grammatically incorrect. You would either have to say "Removes the timeout from a user", which would sound odd since multiple users can be timed out at once, resulting in server moderators having to manage multiple user timeouts. Either that, or put in plural, which would then imply that a user can have multiple timeouts. Therefore, saying "a timeout" is the only way to make a compromise in terms of being grammatical correct and users understanding the purpose of the subcommand, in my humble opinion.

Comment (Ladam0203) [reply src/commands/timeout.ts]: I see your point, I have made the change now :)

File: src/commands/timeout.ts [hunk 8]
```diff
@@ -0,0 +1,162 @@
+//Template: https://github.com/Costpap/CostBot/blob/main/src/commands/kick.ts
+
+//imports
```
Comment (Costpap) [inline src/commands/timeout.ts:3]: I don't see a need for such comments, please remove them in order to be consistent with the rest of the codebase.

```suggestion
```

File: src/commands/timeout.ts (hunk 6 above)
Comment (Costpap) [reply src/commands/timeout.ts]: Just got back from testing the command, and while it does work, I was thinking that maybe the check could be done using `typeof duration !== 'number'` as the if condition? This would be consistent with most other checks in the bot's codebase. I'd send a suggestion as part of this comment, but sadly GitHub doesn't allow one to send suggestions on outdated comments.

File: src/commands/timeout.ts [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+        // Typeguard in order to ensure having access to ChatInputCommand interaction options.
+        if (!interaction.isChatInputCommand()) return;
+
+        if (interaction.inGuild() === false) {
+            return interaction.reply({ content: "❌ I can't execute this command inside DMs!", ephemeral: true });
+        }
+
+        if (!interaction.guild.members.me.permissions.has(PermissionFlagsBits.ModerateMembers)) {
+            return interaction.reply({
+                content: '❌ Sorry, I need the `Timeout Members` in order to execute this command.',
```
Comment (Costpap) [inline src/commands/timeout.ts:51]: Whoops! Someone forgot to put "permission" after `Timeout Members`.
```suggestion
                content: '❌ Sorry, I need the `Timeout Members` permission in order to execute this command.',
```

File: src/commands/timeout.ts [hunk 10]
```diff
@@ -0,0 +1,162 @@
...
+                        ephemeral: true,
+                    });
+                }
+
+                let response = `✅ Timed out \`${member.user.tag} (${member.id})\`.`;
+
+                if (!interaction.options.getBoolean('silent')) {
+                    const notified = await notifyUser(user, interaction, 'timed out');
+                    if (!notified) response += "\n\n⚠️ Couldn't send DM to user.";
+                }
```
Comment (Costpap) [inline src/commands/timeout.ts:111]: This gets sent even if the timeout itself fails. I was thinking maybe we could put it inside the try/catch statement, or below it?

File: src/commands/timeout.ts [hunk 11]
```diff
@@ -0,0 +1,162 @@
...
+                try {
+                    await member.timeout(
+                        null,
+                        interaction.options?.getString('reason') ? `${interaction.options?.getString('reason')}` : '',
+                    );
+                    return interaction.reply({ content: response, ephemeral: true });
+                } catch (error) {
+                    console.error(error);
+                    return interaction.reply({
+                        content: `❌ I encountered an error while trying remove the timeout from \`${
```
Comment (Costpap) [inline src/commands/timeout.ts:153]: Since `timeout` is preceded by a `the` and a user can only have one timeout applied, I feel like using `for` is more natural sounding & grammatically correct than `from`.

```suggestion
                        content: `❌ I encountered an error while trying remove the timeout for \`${
```

File: src/commands/timeout.ts [hunk 12]
```diff
@@ -0,0 +1,162 @@
...
+
+                if (member.moderatable === false) {
+                    return interaction.reply({
+                        content:
+                            '❌ I cannot remove the timeout from this user! \n**Please make sure that my highest role is above theirs.**',
+                        ephemeral: true,
+                    });
+                }
+
+                const response = `✅ Removed the timeout from \`${member.user.tag} (${member.id})\``;
```
Comment (Costpap) [inline src/commands/timeout.ts:142]: Since `timeout` is preceded by a `the` and a user can only have one timeout applied, I feel like using `for` is more natural sounding & grammatically correct than `from`.

```suggestion
                const response = `✅ Removed the timeout for \`${member.user.tag} (${member.id})\``;
```

File: src/commands/timeout.ts [hunk 13]
```diff
@@ -0,0 +1,162 @@
...
+            }
+            case 'remove': {
+                if (!member.isCommunicationDisabled()) {
+                    return interaction.reply({ content: '❌ This user is not currently timed out.', ephemeral: true });
+                }
+
+                if (member.moderatable === false) {
+                    return interaction.reply({
+                        content:
+                            '❌ I cannot remove the timeout from this user! \n**Please make sure that my highest role is above theirs.**',
```
Comment (Costpap) [inline src/commands/timeout.ts:137]: Since `timeout` is preceded by a `the` and a user can only have one timeout applied, I feel like using `for` is more natural sounding & grammatically correct than `from`.

```suggestion
                            '❌ I cannot remove the timeout from this user! \n**Please make sure that my highest role is above theirs.**',
```

**Review by Costpap — CHANGES_REQUESTED** [review]
Hopefully this is the last time I'm requesting changes. I tested pretty much everything I could test, the command works perfectly aside from a few miscellaneous errors here and there, mostly related to strings.

The most important issue I found is that the command doesn't check whether the mute duration is longer than the maximum duration Discord allows a user to be timed out for, which is 28 days, or `2419200000` milliseconds. As soon as this and everything else I've left a comment on is resolved, this pull request should be ready to merge.

Sorry if this got a bit tiring, I barely have spare time nowadays, which makes it harder to deal with any projects I develop outside of ones related to my studies. This is also the first time I'm collaborating with someone who isn't already a close friend of mine. Every other CostBot contributor has been a friend who I've asked or who offered to do something for me. While they were making PRs to submit, I was actively communicating with them through Discord, which is by nature a lot more instant than GitHub review comments. That's not to say contributions like this aren't welcome, they perfectly are, and I do appreciate the time you put into this! That's all for now. :)
